src/Npgsql/ClusterAwareDataSource.cs

```diff
@@ -0,0 +1,489 @@
+                    Debug.Assert(databaseState != DatabaseState.Unknown);
+                    if (!stateValidator(databaseState, preferredType))
+                    {
+                        pool.Return(connector);
+                        return null;
+                    }
+                }
+                return connector;
+            }
+            else
+            {
+                connector = await pool.OpenNewConnector(conn, new NpgsqlTimeout(timeoutPerHost), async, cancellationToken);
+                if (connector is not null)
+                {
+                    if (databaseState == DatabaseState.Unknown)
+                    {
+                        // While opening a new connector we might have refreshed the database state, check again
+                        databaseState = pool.GetDatabaseState();
+                        if (databaseState == DatabaseState.Unknown)
+                            databaseState = await connector.QueryDatabaseState(new NpgsqlTimeout(timeoutPerHost), async, cancellationToken);
+                        Debug.Assert(databaseState != DatabaseState.Unknown);
+                        if (!stateValidator(databaseState, preferredType))
+                        {
+                            pool.Return(connector);
+                            return null;
+                        }
+                    }
+                    return connector;
+                }
+            }
+        }
+        catch (Exception ex)
+        {
+            exceptions.Add(ex);
+            if (connector is not null)
+                pool.Return(connector);
+        }
+
+        return null;
+    }
+    
+    async ValueTask<NpgsqlConnector?> TryGet(
+        NpgsqlConnection conn,
+        TimeSpan timeoutPerHost,
+        bool async,
+        TargetSessionAttributes preferredType,
+        Func<DatabaseState, TargetSessionAttributes, bool> stateValidator,
+        int poolIndex,
+        IList<Exception> exceptions,
+        CancellationToken cancellationToken)
+    {
+        var pools = _pools;
+        var pool = pools[poolIndex];
+        var databaseState = pool.GetDatabaseState();
+        NpgsqlConnector? connector = null;
+    
+        try
+        {
+            connector = await pool.Get(conn, new NpgsqlTimeout(timeoutPerHost), async, cancellationToken);
+            if (databaseState == DatabaseState.Unknown)
+            {
+                // Get might have opened a new physical connection and refreshed the database state, check again
+                databaseState = pool.GetDatabaseState();
+                if (databaseState == DatabaseState.Unknown)
+                    databaseState = await connector.QueryDatabaseState(new NpgsqlTimeout(timeoutPerHost), async, cancellationToken);
+    
+                Debug.Assert(databaseState != DatabaseState.Unknown);
+                if (!stateValidator(databaseState, preferredType))
+                {
+                    pool.Return(connector);
```

Comment:
The `Return()` is not implemented forr `ClusterAwareDataSource`. Is this code unreachable then?